src/widgets/index/index.ts

```diff
@@ -37,6 +37,29 @@ export type Index = Widget & {
   dispose(options: DisposeOptions): void;
 };
 
+function isIndexWidget(widget: Widget): widget is Index {
+  return widget.$$type === 'ais.index';
+}
+
+function resetPageFromIndexWidgetsRecursively(indexWidgets: Index[]): void {
+  if (indexWidgets.length === 0) {
+    return;
+  }
+
+  indexWidgets.forEach(widget => {
+    const widgetHelper = widget.getHelper()!;
+
+    widgetHelper.overrideStateWithoutTriggeringChangeEvent(
+      // @ts-ignore @TODO: remove "ts-ignore" once `resetPage()` is typed in the helper
+      widgetHelper.state.resetPage()
+    );
+
+    resetPageFromIndexWidgetsRecursively(
```

Comment:
It would look something like this, which is not necessarily more readable IMO. Now we don't leverage the single-time index filtering. Besides, we reassign the iterator which might create some inconsistent behaviors. WDYT?

```ts
function resetPageFromWidgets(widgets: Widget[]): void {
  let indexWidgets = widgets.filter(isIndexWidget);

  while (indexWidgets.length > 0) {
    indexWidgets.forEach(widget => {
      const widgetHelper = widget.getHelper()!;

      widgetHelper.overrideStateWithoutTriggeringChangeEvent(
        widgetHelper.state.resetPage()
      );

      indexWidgets = widget.getWidgets().filter(isIndexWidget);
    });
  }
}
```

To [compare with this](https://github.com/algolia/instantsearch.js/blob/063e5bef0e94357f5c5abac6594279680a285bb1/src/widgets/index/index.ts#L44-L61).

Reply:
This implementation does not behave like the recursive one, we only reset the widgets of the last index. We'll extract the "walk" logic at some point I think, it will be useful for URLSync. We can investigate a solution without recursion at that point.